Added logic to consider poison dice in the calculation of the "can still win" condition

blackshadowshade opened the pull request:
Fixes #2040. Fixes #2008. Fixes #1385.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1223/ (if it passes)

cgolubi1 commented:
No one has commented on this proposed change.

cgolubi1 commented:
Okay, so, here's the responderTest in which the site now reports that one player cannot win:

Both players can still win here:
![both_can_still_win](https://cloud.githubusercontent.com/assets/2898207/20735808/8d1b712c-b670-11e6-8476-404697c2b800.png)

Vicious can no longer win here (you don't see it in the UI because the game is from Dead Guy's POV, but it's in the API and the test):
![vicious_cannot_win](https://cloud.githubusercontent.com/assets/2898207/20735822/9fbe38fa-b670-11e6-8b3c-54f59159e10c.png)

So that seems pretty good, and implies that the backend isn't *overcounting* the value of poison dice --- it seems to correctly note that Vicious can't gain more than 22 sides by taking the (10) and losing the p(12), and Vicious is down 22.7 sides.  So that just leaves the question of whether the backend is *undercounting* the value of poison dice.

cgolubi1 commented:
The other three games that are represented in the responderTests are:
* 002: Jellybean vs Dirgo - both players can still win at all points during the game as long as swing dice are set, and i believe this is correct
* 034: bobby 5150 vs Wiseman - both players can still win at all points during the game as long as swing/option dice are set, and i believe this is correct
* 045: bobby 5150 vs Uncle Scratchy - both players can still will and all points during the game in which swing/option dice are set and there are still any poison dice on the board.  There's a late game scenario in which bobby 5150 can no longer win, but this is after all the poison dice have been captured, so the old code got this right already, and it's still right:

![bobby_cant_win](https://cloud.githubusercontent.com/assets/2898207/20735966/91a13730-b671-11e6-8101-47e9655930e0.png)

cgolubi1 commented:
These other games in the corpus also contain Poison dice, and are unchanged.  Does it make sense that they should be?
* 010: Washu vs Hooloovoo: Hooloovoo's rp die is never brought into play, so that makes sense --- win determination always worked here
* 013: wranklepig vs Wiseman: wranklepig's null die also prevents win determination, and is always present when the poison die is
* 020: fendrin vs gman97216: null and mighty dice both prevent win determination
* 030: dexx vs GorgorBey: mood, mighty, and warrior prevent win determination
* 033: RandomBMMixed mirror: null, weak
* 039: wtrollkin vs Maryland: radioactive, rage, morphing

I think all of that makes sense --- except for the first one, it is reasonable that win determination can never be calculated in these games, and it's reasonable that it was always calculable in the first one.

cgolubi1 commented:
Okay, i feel pretty good about this - i added onto test 046 a second round, similar to the first, but instead of capturing 32 sides of dice with the Berserk attack, Dead Guy captures only 30 sides.  Now we wind up with:
![vicious_can_still_win](https://cloud.githubusercontent.com/assets/2898207/20736726/5916706a-b676-11e6-99a1-1baaf2ce32a6.png)

and the API says both players can still win.  This seems like pretty good evidence that the math is correctly valuing poison dice.  Here's the diff with my test addition:
```diff
diff --git a/test/src/api/responderTest.php b/test/src/api/responderTest.php
index a0dd167..c872a7b 100644
--- a/test/src/api/responderTest.php
+++ b/test/src/api/responderTest.php
@@ -14598,6 +14598,7 @@ class responderTest extends PHPUnit_Framework_TestCase {
 
     /**
      * @depends test_request_savePlayerInfo
+     * @group fulltest_deps
      *
      * This is a basic test of a game containing Dead Guy, to be used in
      * regression-testing Dead Guy behavior
@@ -14755,6 +14756,134 @@ class responderTest extends PHPUnit_Framework_TestCase {
         $expData['playerDataArray'][1]['canStillWin'] = false;
 
         $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+
+
+        ////////////////////
+        // Move 05 - responder004 surrenders
+        $_SESSION = $this->mock_test_user_login('responder004');
+        $this->verify_api_submitTurn(
+            array(0, 0, 0, 0, 0, 13, 4, 2, 6, 5, 7),
+            'responder004 surrendered. End of round: responder003 won round 1 because opponent surrendered. ',
+            $retval, array(),
+            $gameId, 1, 'Surrender', 1, 0, '');
+        $_SESSION = $this->mock_test_user_login('responder003');
+
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder004', 'message' => 'responder004 surrendered
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder003', 'message' => 'End of round: responder0
+        $expData['gameActionLogCount'] = 7;
+        $expData['playerDataArray'][0]['activeDieArray'] = array(
+            array('value' => NULL, 'sides' => 0, 'skills' => array(), 'properties' => array(), 'recipe' => '(0)', 'description' => '0-sided die'),
+            array('value' => NULL, 'sides' => 0, 'skills' => array(), 'properties' => array(), 'recipe' => '(0)', 'description' => '0-sided die'),
+            array('value' => NULL, 'sides' => 0, 'skills' => array(), 'properties' => array(), 'recipe' => '(0)', 'description' => '0-sided die'),
+            array('value' => NULL, 'sides' => 0, 'skills' => array(), 'properties' => array(), 'recipe' => '(0)', 'description' => '0-sided die'),
+            array('value' => NULL, 'sides' => 0, 'skills' => array(), 'properties' => array(), 'recipe' => '(0)', 'description' => '0-sided die'),
+            array('value' => NULL, 'sides' => 20, 'skills' => array('Berserk'), 'properties' => array(), 'recipe' => 'B(20)', 'description' => 'Berse
+        );
+        $expData['playerDataArray'][1]['activeDieArray'] = array(
+            array('value' => NULL, 'sides' => 4, 'skills' => array(), 'properties' => array(), 'recipe' => '(4)', 'description' => '4-sided die'),
+            array('value' => NULL, 'sides' => 6, 'skills' => array(), 'properties' => array(), 'recipe' => '(6)', 'description' => '6-sided die'),
+            array('value' => NULL, 'sides' => 12, 'skills' => array('Shadow'), 'properties' => array(), 'recipe' => 's(12)', 'description' => 'Shadow
+            array('value' => NULL, 'sides' => 12, 'skills' => array('Poison'), 'properties' => array(), 'recipe' => 'p(12)', 'description' => 'Poison
+            array('value' => NULL, 'sides' => NULL, 'skills' => array(), 'properties' => array(), 'recipe' => '(X)', 'description' => 'X Swing Die'),
+            array('value' => NULL, 'sides' => 20, 'skills' => array('Berserk'), 'properties' => array(), 'recipe' => 'B(20)', 'description' => 'Berse
+        );
+        $expData['playerDataArray'][0]['capturedDieArray'] = array();
+        $expData['playerDataArray'][1]['capturedDieArray'] = array();
+        $expData['playerDataArray'][0]['canStillWin'] = NULL;
+        $expData['playerDataArray'][1]['canStillWin'] = NULL;
+        $expData['playerDataArray'][0]['gameScoreArray']['W'] = 1;
+        $expData['playerDataArray'][1]['gameScoreArray']['L'] = 1;
+        $expData['playerDataArray'][0]['roundScore'] = NULL;
+        $expData['playerDataArray'][0]['sideScore'] = NULL;
+        $expData['playerDataArray'][1]['roundScore'] = NULL;
+        $expData['playerDataArray'][1]['sideScore'] = NULL;
+        $expData['playerDataArray'][1]['swingRequestArray'] = array('X' => array(4, 20));
+        $expData['playerDataArray'][1]['prevSwingValueArray'] = array('X' => 20);
+        $expData['validAttackTypeArray'] = array();
+        $expData['gameState'] = 'SPECIFY_DICE';
+        $expData['activePlayerIdx'] = NULL;
+        $expData['roundNumber'] = 2;
+
+        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+
+
+        ////////////////////
+        // Move 06 - responder004 submits die values
+        $_SESSION = $this->mock_test_user_login('responder004');
+        $this->verify_api_submitDieValues(
+            array(0, 0, 0, 0, 0, 15), // BUG #2041: Dead Guy's dice shouldn't be rerolled
+            $gameId, 2, array('X' => 20), NULL);
+        $_SESSION = $this->mock_test_user_login('responder003');
+
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder004', 'message' => 'responder004 set swing v
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => '', 'message' => 'responder003 won initiative for roun
+        $expData['gameActionLogCount'] = 9;
+        $expData['gameState'] = 'START_TURN';
+        $expData['activePlayerIdx'] = 0;
+        $expData['playerWithInitiativeIdx'] = 0;
+        $expData['validAttackTypeArray'] = array('Power', 'Berserk');
+        $expData['playerDataArray'][0]['waitingOnAction'] = TRUE;
+        $expData['playerDataArray'][1]['waitingOnAction'] = FALSE;
+        $expData['playerDataArray'][0]['roundScore'] = 10;
+        $expData['playerDataArray'][0]['sideScore'] = -6.0;
+        $expData['playerDataArray'][1]['roundScore'] = 19;
+        $expData['playerDataArray'][1]['sideScore'] = 6.0;
+        $expData['playerDataArray'][1]['swingRequestArray'] = array();
+        $expData['playerDataArray'][0]['activeDieArray'][0]['value'] = 0;
+        $expData['playerDataArray'][0]['activeDieArray'][1]['value'] = 0;
+        $expData['playerDataArray'][0]['activeDieArray'][2]['value'] = 0;
+        $expData['playerDataArray'][0]['activeDieArray'][3]['value'] = 0;
+        $expData['playerDataArray'][0]['activeDieArray'][4]['value'] = 0;
+        $expData['playerDataArray'][0]['activeDieArray'][5]['value'] = 13;
+        $expData['playerDataArray'][1]['activeDieArray'][0]['value'] = 4;
+        $expData['playerDataArray'][1]['activeDieArray'][1]['value'] = 2;
+        $expData['playerDataArray'][1]['activeDieArray'][2]['value'] = 6;
+        $expData['playerDataArray'][1]['activeDieArray'][3]['value'] = 5;
+        $expData['playerDataArray'][1]['activeDieArray'][4]['value'] = 15;
+        $expData['playerDataArray'][1]['activeDieArray'][4]['sides'] = 20;
+        $expData['playerDataArray'][1]['activeDieArray'][4]['description'] = 'X Swing Die (with 20 sides)';
+        $expData['playerDataArray'][1]['activeDieArray'][5]['value'] = 7;
+        $expData['playerDataArray'][1]['prevSwingValueArray'] = array();
+
+        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+
+
+        ////////////////////
+        // Move 07 - responder003 performed Berserk attack using [B(20):13] against [(4):4,(6):2,B(20):7]
+        $this->verify_api_submitTurn(
+            array(6),
+            'responder003 performed Berserk attack using [B(20):13] against [(4):4,(6):2,B(20):7]; Defender (4) was captured; Defender (6) was captur
+            $retval, array(array(0, 5), array(1, 0), array(1, 1), array(1, 5)),
+            $gameId, 2, 'Berserk', 0, 1, '');
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder003', 'message' => 'responder003 performed B
+        $expData['gameActionLogCount'] = 10;
+        $expData['activePlayerIdx'] = 1;
+        $expData['validAttackTypeArray'] = array('Power', 'Skill', 'Shadow');
+        $expData['playerDataArray'][0]['roundScore'] = 35;
+        $expData['playerDataArray'][1]['roundScore'] = 4;
+        $expData['playerDataArray'][0]['sideScore'] = 20.7;
+        $expData['playerDataArray'][1]['sideScore'] = -20.7;
+        $expData['playerDataArray'][0]['waitingOnAction'] = FALSE;
+        $expData['playerDataArray'][1]['waitingOnAction'] = TRUE;
+        $expData['playerDataArray'][0]['activeDieArray'][5]['value'] = 6;
+        $expData['playerDataArray'][0]['activeDieArray'][5]['sides'] = 10;
+        $expData['playerDataArray'][0]['activeDieArray'][5]['recipe'] = '(10)';
+        $expData['playerDataArray'][0]['activeDieArray'][5]['description'] = '10-sided die';
+        $expData['playerDataArray'][0]['activeDieArray'][5]['properties'] = array('HasJustSplit', 'JustPerformedBerserkAttack');
+        $expData['playerDataArray'][0]['activeDieArray'][5]['skills'] = array();
+        array_splice($expData['playerDataArray'][1]['activeDieArray'], 5, 1);
+        array_splice($expData['playerDataArray'][1]['activeDieArray'], 0, 2);
+        $expData['playerDataArray'][0]['capturedDieArray'] = array(
+            array('value' => 4, 'sides' => 4, 'recipe' => '(4)', 'properties' => array('WasJustCaptured')),
+            array('value' => 2, 'sides' => 6, 'recipe' => '(6)', 'properties' => array('WasJustCaptured')),
+            array('value' => 7, 'sides' => 20, 'recipe' => 'B(20)', 'properties' => array('WasJustCaptured')),
+        );
+        $expData['playerDataArray'][0]['canStillWin'] = TRUE;
+        $expData['playerDataArray'][1]['canStillWin'] = TRUE;
+
+        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
+
+
     }
 
     /**
```

cgolubi1 commented:
So, if you integrate that addition into test 046, that will leave me fairly happy that we are in good shape here.

One note is that the stuff i added includes another triggering of #2083 so you'll have to sort that out merge-wise.

irilyth commented:
That all looks good to me. Shipit!

blackshadowshade commented:
Okay, added the changes required to the responder tests. Testing pull request: #2141.

blackshadowshade commented:
Yeah, I don't know what to do about this pull request. It obviously fails Jenkins because it's dependent on #2083. Do you want #2083 merged in here, and then the pull request refreshed once it's in master?

cgolubi1 commented:
So... no, i don't want #2083 merged into this pull.  That seems like the wrong type of answer.

cgolubi1 commented:
So... the diff i actually gave you had the unnecessary 0 rerolls, and (i believe) would have passed Jenkins.  You took those zeroes out, right?

In the short-term, i think the right thing to do is:
* Put those zeroes back in.  Those zeroes cause this pull to be consistent with the state of master, and thus cause Jenkins to pass.
* When we merge both #2083 and this pull, we'll discover that Jenkins no longer passes, and we'll have to do a pull of changes to make everything happy.

blackshadowshade commented:
Okay, I'll put those back in then.

cgolubi1 commented:
In the long-term, i think this is a good demonstration in a nutshell of why the new world order has problems, and i have some ideas about that.  Happy to talk about it on IRC for half an hour now, or send e-mail.

cgolubi1 commented:
As noted on #2133 punting this for December until the new workflow is sorted.

cgolubi1 commented:
It looks like people agreed that this was a good change.  Rebase it down to one commit, then i'll throw some replay tests at it, and we can put it over the fence.

blackshadowshade commented:
http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1287/ (if it passes)

blackshadowshade commented:
Rebased to be after the turbo commit.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1310/ (if it passes)

blackshadowshade commented:
So, Jenkins fails on this for the obvious reason, and I think I'll hold off making this work until #2083 has been accepted.

blackshadowshade commented:
I've fixed the unit tests, now that #2083 has been accepted, and I've rebased to one commit.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1353/ (if it passes)

cgolubi1 commented on test/src/api/responderTest.php:
```diff
@@ -14923,10 +14953,137 @@ public function test_interface_game_046() {
+
+
+        ////////////////////
+        // Move 05 - responder004 surrenders
```
Why did you add moves 5-7 to this test?  Is this a merge artifact, or are these moves relevant to "can still win" determination testing?

blackshadowshade replied:
I can't remember exactly, but it could have had to do with requiring the canStillWin to be NULL at game states outside of the round. Or perhaps it was something to do with Berserk and canStillWin. I'm happy to remove these unit tests if you'd like.

blackshadowshade replied:
On second thought, after looking at #2108, it's quite possible that it is a merge conflict gone wrong.

blackshadowshade replied:
No wait! This responder test comes directly from your post on 30 Nov 2016, which you asked me to add yourself.  :)

cgolubi1 replied:
Ah, yeah, okay.  Indeed.

cgolubi1 commented:
I'll want to do some replay testing on this, and that's going to take a bit of doing, so i'll try to get that engine running.

Sooner than that, i'll stand up a dev site for this, probably tomorrow.

cgolubi1 commented:
Dev site is up, at 54.224.87.125

blackshadowshade commented:
Just reminding testers that we need to do some manual testing on this. Check your stale games.

blackshadowshade commented:
I've rebased this to deal with conflicts in the responder tests.

blackshadowshade commented:
So, manual testing is still ongoing on this, but so far, things look good.

cgolubi1 commented:
This has been outstanding for a long time.  Shadowshade, any reason we can't merge it?

blackshadowshade commented:
I reckon this is good to go.
